### PR TITLE
Expose `OpaqueIpcMessage` as `IpcMessage`

### PR DESCRIPTION
**Pull request by mrobinson**
This makes the API a bit friendlier and future-proof. Instead of
returning a tuple from raw `recv()` return the `IpcMessage` struct that
is used internally in the crate. In addition, this is used to pass data
from the platform layer -- removing clippy warnings about complex data
types. One upside to this is that `Option<IpcSharedMemory>` is turned
into `IpcSharedMemory` in some places where having a `None` value
shouldn't be possible.

This simplification removes about 300 lines of code.

### CODE REVIEW

**Comment by mrobinson** [comment]
This will need #364 to work properly.

File: Cargo.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ name = "ipc_receiver_set"
 harness = false
 
 [features]
+default = ["async"]
```
Comment (jdm) [inline Cargo.toml:23]: Why this change?

File: src/ipc.rs [hunk 2]
```diff
@@ -735,14 +724,17 @@ impl OpaqueIpcMessage {
...
+                            .into_iter()
+                            .map(Some)
+                            .collect(),
                     );
                     let result = bincode::deserialize(&self.data[..]);
-                    mem::swap(
+                    let _ = mem::replace(
                         &mut *os_ipc_shared_memory_regions_for_deserialization.borrow_mut(),
-                        &mut self.os_ipc_shared_memory_regions,
+                        old_ipc_shared_memory_regions_for_deserialization,
```
Comment (jdm) [inline src/ipc.rs:737]: I think this would be equivalent to (and more clear as):
```rust
*os_ipc_shared_memory_regions_for_deserialization.borrow_mut() = old_ipc_shared_memory_regions_for_deserialization;
```

**Review by jdm — APPROVED** [review]
This is a really nice reduction in complexity!

File: src/ipc.rs (hunk 2 above)
Comment (mrobinson) [reply src/ipc.rs]: :+1:

**Comment by mrobinson** [comment]
Thanks for the review!

File: src/ipc.rs [hunk 3]
```diff
@@ -699,19 +691,16 @@ impl Debug for OpaqueIpcMessage {
     }
 }
 
-impl OpaqueIpcMessage {
-    fn new(
+impl IpcMessage {
+    pub(crate) fn new(
```
Comment (mukilan) [inline src/ipc.rs:695]: I'm curious if there is reason for not marking this as `pub`. The struct only has pub fields, so it doesn't seem like we are gaining much in terms of encapsulation.

Comment (mrobinson) [reply src/ipc.rs:695]: Fair enough. I think mainly this wasn't marked as `pub` because it isn't expected that a user of `ipc-channel` would ever need to create this data structure. `pub(crate)` is useful in this case because the compiler can determine if the function ever becomes unused.

Comment (mukilan) [reply src/ipc.rs:695]: > because it isn't expected that a user of ipc-channel would ever need to create this data structure

Should we then remove the 'pub' from the fields? Having them public means it will become part of our API in the next release and OpaqueIpcMessage only [had private fields](https://docs.rs/ipc-channel/latest/ipc_channel/ipc/struct.OpaqueIpcMessage.html). I'm happy to do the change.

Comment (mrobinson) [reply src/ipc.rs:695]: I am not opposed to making the fields private again.
